File: Tests/DrawingTests/DrawingStoreTests.swift
import CoreGraphics
import Foundation
import Testing
@testable import Scratchpad

private final class MockDrawingRepository: DrawingRepository {
    var documentsByDisplay: [DisplayIdentifier: DrawingDocument] = [:]
    private(set) var saveCount = 0

    func load(for display: DisplayIdentifier) -> DrawingDocument {
        documentsByDisplay[display] ?? .empty
    }

    func save(_ document: DrawingDocument, for display: DisplayIdentifier) throws {
        saveCount += 1
        documentsByDisplay[display] = document
    }

    func clear(for display: DisplayIdentifier) throws {
        documentsByDisplay.removeValue(forKey: display)
    }
}

struct DrawingStoreTests {

    private let display = DisplayIdentifier(rawValue: 1)

    @Test func loadsPersistedStrokesOnInit() {
        let repository = MockDrawingRepository()
        let stroke = Stroke(points: [StrokePoint(x: 0, y: 0)])
        repository.documentsByDisplay[display] = DrawingDocument(strokes: [stroke], updatedAt: Date())

        let store = DrawingStore(display: display, repository: repository)

        #expect(store.strokes == [stroke])
    }

    @Test func beginAppendEndBuildsACompletedStroke() {
        let store = DrawingStore(display: display, repository: MockDrawingRepository())

        store.beginStroke(at: CGPoint(x: 0, y: 0))
        store.appendPoint(CGPoint(x: 5, y: 5))
        store.appendPoint(CGPoint(x: 10, y: 10))
        store.endStroke()

        #expect(store.strokes.count == 1)
        #expect(store.strokes[0].points.map(\.cgPoint) == [
            CGPoint(x: 0, y: 0), CGPoint(x: 5, y: 5), CGPoint(x: 10, y: 10),
        ])
        #expect(store.strokeInProgress == nil)
    }

    @Test func endStrokeWithoutBeginDoesNothing() {
        let store = DrawingStore(display: display, repository: MockDrawingRepository())
        store.endStroke()
        #expect(store.strokes.isEmpty)
    }

    @Test func clearRemovesAllStrokesAndAnyInProgressStroke() {
        let store = DrawingStore(display: display, repository: MockDrawingRepository())
        store.beginStroke(at: CGPoint(x: 0, y: 0))
        store.appendPoint(CGPoint(x: 1, y: 1))
        store.endStroke()
        store.beginStroke(at: CGPoint(x: 2, y: 2))

        store.clear()

        #expect(store.strokes.isEmpty)
        #expect(store.strokeInProgress == nil)
    }
}

File: Windows/Scratchpad.Tests/DrawingTests/DrawingStoreTests.cs
using Scratchpad.Drawing;
using Scratchpad.Overlay;
using Scratchpad.Persistence;
using Xunit;

namespace Scratchpad.Tests.DrawingTests;

internal sealed class MockDrawingRepository : IDrawingRepository
{
    public Dictionary<DisplayIdentifier, DrawingDocument> DocumentsByDisplay { get; } = new();
    public int SaveCount { get; private set; }

    public DrawingDocument Load(DisplayIdentifier display) =>
        DocumentsByDisplay.TryGetValue(display, out var document) ? document : DrawingDocument.Empty;

    public void Save(DrawingDocument document, DisplayIdentifier display)
    {
        SaveCount++;
        DocumentsByDisplay[display] = document;
    }

    public void Clear(DisplayIdentifier display) => DocumentsByDisplay.Remove(display);
}

public class DrawingStoreTests
{
    private static readonly DisplayIdentifier Display = new("\\.\\DISPLAY-TEST");

    [Fact]
    public void LoadsPersistedStrokesOnInit()
    {
        var repository = new MockDrawingRepository();
        var stroke = new Stroke { Points = { new StrokePoint(0, 0) } };
        repository.DocumentsByDisplay[Display] = new DrawingDocument { Strokes = { stroke } };

        var store = new DrawingStore(Display, repository);

        Assert.Single(store.Strokes);
        Assert.Equal(stroke.Id, store.Strokes[0].Id);
    }

    [Fact]
    public void BeginAppendEndBuildsACompletedStroke()
    {
        var store = new DrawingStore(Display, new MockDrawingRepository());

        store.BeginStroke(new StrokePoint(0, 0));
        store.AppendPoint(new StrokePoint(5, 5));
        store.AppendPoint(new StrokePoint(10, 10));
        store.EndStroke();

        Assert.Single(store.Strokes);
        Assert.Equal(
            new[] { new StrokePoint(0, 0), new StrokePoint(5, 5), new StrokePoint(10, 10) },
            store.Strokes[0].Points);
        Assert.Null(store.StrokeInProgress);
    }

    [Fact]
    public void EndStrokeWithoutBeginDoesNothing()
    {
        var store = new DrawingStore(Display, new MockDrawingRepository());
        store.EndStroke();
        Assert.Empty(store.Strokes);
    }

    [Fact]
    public void ClearRemovesAllStrokesAndAnyInProgressStroke()
    {
        var store = new DrawingStore(Display, new MockDrawingRepository());
        store.BeginStroke(new StrokePoint(0, 0));
        store.AppendPoint(new StrokePoint(1, 1));
        store.EndStroke();
        store.BeginStroke(new StrokePoint(2, 2));

        store.Clear();

        Assert.Empty(store.Strokes);
        Assert.Null(store.StrokeInProgress);
    }

    [Fact]
    public void RemoveStrokesTouchingErasesTheWholeStrokeItTouches()
    {
        var store = new DrawingStore(Display, new MockDrawingRepository());
        store.BeginStroke(new StrokePoint(0, 0));
        store.AppendPoint(new StrokePoint(10, 0));
        store.EndStroke();

        var erasedIds = store.RemoveStrokes(new StrokePoint(5, 0), tolerance: 2);

        Assert.Single(erasedIds);
        Assert.Empty(store.Strokes);
    }

    [Fact]
    public void RemoveStrokesMissesAPointFarFromAnyStroke()
    {
        var store = new DrawingStore(Display, new MockDrawingRepository());
        store.BeginStroke(new StrokePoint(0, 0));
        store.AppendPoint(new StrokePoint(10, 0));
        store.EndStroke();

        var erasedIds = store.RemoveStrokes(new StrokePoint(0, 100), tolerance: 2);

        Assert.Empty(erasedIds);
        Assert.Single(store.Strokes);
    }
}
